src/main/scala/dpla/ingestion3/mappers/providers/HathiMapping.scala

```diff
@@ -0,0 +1,703 @@
+    "pd" -> "Public domain",
+    "ic-world" -> "In-copyright and permitted as world viewable by the copyright holder",
+    "pdus" -> "Public domain only when viewed in the US",
+    "cc-by" -> "Creative Commons Attribution license",
+    "cc-by-nd" -> "Creative Commons Attribution-NoDerivatives license",
+    "cc-by-nc-nd" -> "Creative Commons Attribution-NonCommercial-NoDerivatives license",
+    "cc-by-nc" -> "Creative Commons Attribution-NonCommercial license",
+    "cc-by-nc-sa" -> "Creative Commons Attribution-NonCommercial-ShareAlike license",
+    "cc-by-sa" -> "Creative Commons Attribution-ShareAlike license",
+    "cc-zero" -> "Creative Commons Zero license (implies pd)",
+    "und-world" -> "undetermined copyright status and permitted as world viewable by the depositor"
+  )
+
+  private val googlePrefixMapping: Map[String, String] = Map(
+    "chi" -> "CHI",
+    "coo" -> "CORNELL",
+    "hvd" -> "HARVARD",
+    "ien" -> "NWU",
+    "inu" -> "IND",
+    "mdp" -> "UOM",
+    "nnc1" -> "COLUMBIA",
+    "nyp" -> "NYPL",
+    "pst" -> "PSU",
+    "pur1" -> "PURD",
+    "uc1" -> "UCAL",
+    "ucm" -> "UCM",
+    "umn" -> "MINN",
+    "uva" -> "UVA",
+    "wu" -> "WISC"
+  )
+
+  private def getUcalPrefix(barcode: String): String = {
+    if (barcode.length == 11 && barcode.startsWith("l")) "UCLA"
+    else if (barcode.length == 10) "UCB"
+    else if (barcode.length == 14) {
+      barcode.slice(1,5) match {
+        case "1822" => "UCSD"
+        case "1970" => "UCI"
+        case "1378" => "UCSF"
+        case "2106" => "UCSC"
+        case "1205" => "UCSB"
+        case "1175" => "UCD"
+        case "1158" => "UCLA"
+        case "1210" => "UCR"
+        case _ => "UCAL"
+      }
+    } else "UCAL"
+  }
+}
+
+class HathiThumbnailFetcher(hathiIdOpt: Option[String],
+                            oclcIdOpt: Option[String],
+                            isbnIdOpt: Option[String],
+                            googlePrefixOpt: Option[String]) extends JsonExtractor {
+
+  val baseUrl: String = "http://books.google.com/books?jscmd=viewapi&bibkeys="
+
+  val hathiId: String = hathiIdOpt.getOrElse("")
+  val oclcId: String = oclcIdOpt.getOrElse("")
+  val isbnId: String = isbnIdOpt.getOrElse("")
+  val googlePrefix: String = googlePrefixOpt.getOrElse("")
+
+  // Thumbnail URL request:
+  // Base URL: http://books.google.com/books?jscmd=viewapi&bibkeys=<params>
+  // Where params are: <google_prefix>:<hathi_id>,OCLC<oclc_id>,ISBN:<isbn>
+  // ISBN ID is optional
+  val requestUrl: Option[String] = {
+    if (hathiId.isEmpty || oclcId.isEmpty || googlePrefix.isEmpty) None
+    else {
+      val isbnSuffix = if (isbnId.isEmpty) "" else ",ISBN:" + isbnId
+      val url = baseUrl + googlePrefix + ":" + hathiId + ",OCLC:" + oclcId + isbnSuffix
+      Some(url)
+    }
+  }
+
+  // Control flow for entire process of constructing request URL, sending request, and processing response.
+  val thumbnailUrl: Option[String] = {
```

Comment:
It might be cleaner to chain calls to `.flatMap` for this.

Reply:
good idea